File: 637. Average of Levels in Binary Tree/main.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        result = list()

        if not root:
            return result

        q = list()
        q.append(root)

        while q:
            size = len(q)
            value = 0

            for _ in range(size):
                node = q.pop(0)
                value += node.val

                if node.left:
                    q.append(node.left)

                if node.right:
                    q.append(node.right)

            result.append(value / size)

        return result


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        info = list()

        def dfs(node: Optional[TreeNode], level: int):
            if not node:
                return

            if len(info) <= level:
                info.append([0, 0])

            info[level][0] += node.val
            info[level][1] += 1

            dfs(node.left, level + 1)
            dfs(node.right, level + 1)

        dfs(root, 0)

        return [value / count for value, count in info]


from collections import deque


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        result = list()

        q = deque()
        q.append(root)

        while q:
            size = len(q)
            value = 0

            for _ in range(size):
                node = q.popleft()
                value += node.val

                if node.left:
                    q.append(node.left)

                if node.right:
                    q.append(node.right)

            result.append(value / size)

        return result

File: 637. Average of Levels in Binary Tree/main.rs
struct Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn average_of_levels(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<f64> {
        let mut result = Vec::new();
        if root.is_none() {
            return result;
        }

        let mut q = Vec::new();

        q.push(root);

        while !q.is_empty() {
            let size = q.len();
            let mut value = 0.0;

            for _i in 0..size {
                if let Some(node) = q.remove(0) {
                    let node_ref = node.borrow();
                    value += node_ref.val as f64;

                    if node_ref.left.is_some() {
                        q.push(node_ref.left.clone());
                    }

                    if node_ref.right.is_some() {
                        q.push(node_ref.right.clone());
                    }
                }
            }

            result.push(value / size as f64);
        }

        result
    }
}
